protocol/auditlog.go

```diff
@@ -0,0 +1,214 @@
+// This module implements a CONIKS audit log that a CONIKS auditor
+// maintains.
+// An audit log is a mirror of many CONIKS key directories' STR history,
+// allowing CONIKS clients to audit the CONIKS directories.
+
+package protocol
+
+import (
+	"encoding/hex"
+
+	"github.com/coniks-sys/coniks-go/crypto"
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+)
+
+type directoryHistory struct {
+	dirName   string
+	signKey   sign.PublicKey
+	snapshots map[uint64]*DirSTR
+	latestSTR *DirSTR
+}
+
+// A ConiksAuditLog maintains the histories
+// of all CONIKS directories known to a CONIKS auditor,
+// indexing the histories by the hash of a directory's initial
+// STR (specifically, the hash of the STR's signature as a string).
+// Each history includes the directory's domain name as a string, its
+// public signing key enabling the auditor to verify the corresponding
+// signed tree roots, and a map with the snapshots for each observed
+// epoch.
+type ConiksAuditLog map[string]*directoryHistory
+
+// computeInitSTRHash is a wrapper for the digest function;
+// returns empty string if the STR isn't an initial STR (i.e. str.Epoch != 0)
+func computeInitSTRHash(initSTR *DirSTR) string {
+	if initSTR.Epoch != 0 {
```

Comment:
I refactored this function in https://github.com/coniks-sys/coniks-go/pull/155/commits/92cee27f21ab8299483b4d43712cfbd05c8308a6 (it would be used by the clients, too). 
Feel free to revert if you don't like it.